### DISCUSSION TITLE
WIdgets in Niri

### DISCUSSION THREAD

**Discussion by aditya23043** [YaLTeR/niri#324] [Q&A] [answered]
Just like AwesomeWM, I wanted to make basic widgets for stuff like volume, brightness etc, and I have noticed that niri has some widgets like this in the form of the hotkey menu and the quit confirmation box. I am not sure if that's niri or a different program.
If niri has support for basic widgets, can someone provide the code for how the exit confirmation dialogue works?
Thanks!

**Comment by YaLTeR** [comment, marked as answer]
Hey, these two are indeed in niri, but they are essentially 100% "hardcoded". In order to keep things simple, I didn't want to add a whole GUI toolkit to niri, so these dialogs are all manually laid out and drawn with Cairo, with quite a bit of boilerplate. You can find the code here: https://github.com/YaLTeR/niri/tree/main/src/ui

**Reply by YaLTeR** [reply]
Besides, volume and brightness tracking in itself is likely nontrivial, so I suggest for now using some layer-shell widget framework for this.

**Reply by aditya23043** [reply]
got it thank you!
